source/s3_client.c

```diff
@@ -930,253 +909,179 @@ static void s_s3_client_process_work_task(struct aws_task *task, void *arg, enum
     struct aws_s3_client *client = arg;
     AWS_PRECONDITION(client);
 
-    /* Client keeps a reference to the event loop group; a 'canceled' status should not be happen.*/
+    /* Client keeps a reference to the event loop group; a 'canceled' status should not happen.*/
     AWS_ASSERT(task_status == AWS_TASK_STATUS_RUN_READY);
 
-    struct aws_linked_list vip_connections_updates;
-    struct aws_linked_list vip_connections_removals;
-    struct aws_linked_list meta_requests_adds;
-    struct aws_linked_list work_destroy_list;
+    struct aws_linked_list meta_request_work_list;
+    aws_linked_list_init(&meta_request_work_list);
 
+    struct aws_linked_list vip_connections_updates;
     aws_linked_list_init(&vip_connections_updates);
-    aws_linked_list_init(&vip_connections_removals);
-    aws_linked_list_init(&meta_requests_adds);
-    aws_linked_list_init(&work_destroy_list);
 
+    /*******************/
+    /* Step 1: Move everything into thread local memory. */
+    /*******************/
     s_s3_client_lock_synced_data(client);
 
     /* Once we exit this mutex, someone can reschedule this task. */
     client->synced_data.process_work_task_scheduled = false;
 
+    aws_linked_list_swap_contents(&meta_request_work_list, &client->synced_data.pending_meta_request_work);
```

Comment:
Should we add a precondition that the `pending_meta_request_work` list is empty?

Reply:
I don't think so?  We're trying to grab everything in pending_meta_requests_work so that we can use it in the thread.